File: linked_list/21_merge_two_sorted_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Approach:
            - iterate while list1 and list2 curr ptrs are not none
            - keep appending the smallest node to the list
            - append the remainding nodes in one of the non-empty lists
        """
        merged_list_head = None
        list_curr = None
        list1_ptr, list2_ptr = list1, list2

        while list1_ptr and list2_ptr:
            new_node = None
            if list1_ptr.val < list2_ptr.val:
                new_node = ListNode(list1_ptr.val)
                list1_ptr = list1_ptr.next
            else:
                new_node = ListNode(list2_ptr.val)
                list2_ptr = list2_ptr.next
            
            if merged_list_head is None:
                merged_list_head = new_node
                list_curr = merged_list_head
            else:
                list_curr.next = new_node
                list_curr = list_curr.next
        
        if list1_ptr and merged_list_head:
            # remaining list 1 nodes exist, append
            list_curr.next = list1_ptr
        else: merged_list_head = list1_ptr
        
        if list2_ptr and merged_list_head:
            # remaining list 2 nodes exist, append
            list_curr.next = list2_ptr
        else: merged_list_head = list2_ptr
            
        return merged_list_head

File: linked_list/21.merge-two-sorted-lists.cpp
/*
 * @lc app=leetcode id=21 lang=cpp
 *
 * [21] Merge Two Sorted Lists
 */

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    /**
     * Approaches:
     * 1. Store the nodes in a buffer and sort then link the nodes (O(nlogn) time, O(n) space)
     * 2. Iterate the two-pointers, maintain output pointer, detach and link nodes accordingly (O(n) time, O(1) space)
     * 
     * I like 2, let's go with that.
     * 
     * Steps:
     * 
     * initialization:
     *      - list1 and list2
     *      - output: merged_head, merged_curr
     * - loop list1 and list2 until either cursor is NULL, selectively adding the nodes to 
     * merged_list based on value
     * - merge remaining non-null list
     */
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2)
    {
        ListNode *merged_head, *merged_curr;

        // nullcase
        if (list1 == nullptr && list2 == nullptr)
        {
            return nullptr;
        }
        else if (list1 == nullptr)
        {
            return list2;
        }
        else if (list2 == nullptr)
        {
            return list1;
        }

        // Init
        merged_head = nullptr;
        merged_curr = nullptr;

        if (list1->val < list2->val)
        {
            merged_head = list1;
            list1 = list1->next;
        }
        else
        {
            merged_head = list2;
            list2 = list2->next;
        }

        merged_curr = merged_head;
        merged_curr->next = nullptr;

        // loop
        while (list1 && list2)
        {
            if (list1->val < list2->val)
            {
                merged_curr->next = list1;
                list1 = list1->next;
            }
            else
            {
                merged_curr->next = list2;
                list2 = list2->next;
            }

            // Increment curr and detach added node from its original list
            merged_curr = merged_curr->next;
            merged_curr->next = nullptr;
        }

        // Update output list with remaining elements
        if (list1 != nullptr)
        {
            merged_curr->next = list1;
        }
        else if (list2 != nullptr)
        {
            merged_curr->next = list2;
        }

        return merged_head;
    }
};
// @lc code=end
